test/mpi/dd_matrix.cpp

```diff
@@ -0,0 +1,596 @@
+                 {5, 0, -0.5}, {5, 2, -1}, {5, 4, -0.5}, {5, 5, 2}}}}},
+          engine(42)
+    {
+        part = part_type::build_from_contiguous(
+            exec, gko::array<global_index_type>(
+                      exec, I<global_index_type>{0, 4, 8, 12}));
+
+        dist_mat = dd_mtx_type::create(exec, comm);
+        x = dist_vec_type::create(ref, comm);
+        y = dist_vec_type::create(ref, comm);
+    }
+
+    void SetUp() override { ASSERT_EQ(comm.size(), 3); }
+
+
+    gko::dim<2> size;
+    gko::dim<2> local_size;
+    std::shared_ptr<part_type> part;
+
+    gko::matrix_data<value_type, global_index_type> mat_input;
+    std::array<matrix_data, 3> dist_input;
+    std::array<local_matrix_data, 3> local_result;
+
+    std::unique_ptr<dd_mtx_type> dist_mat;
+
+    std::unique_ptr<dist_vec_type> x;
+    std::unique_ptr<dist_vec_type> y;
+
+    std::default_random_engine engine;
+};
+
+TYPED_TEST_SUITE(DdMatrix, gko::test::ValueLocalGlobalIndexTypes,
+                 TupleTypenameNameGenerator);
+
+
+TYPED_TEST(DdMatrix, ReadsDistributed)
+{
+    using value_type = typename TestFixture::value_type;
+    using csr = typename TestFixture::local_matrix_type;
+    auto rank = this->comm.rank();
+    auto res_local = csr::create(this->exec);
+    res_local->read(this->local_result[rank]);
+    I<I<value_type>> local_restriction = {{1, 0, 0, 0}, {0, 1, 0, 0},
+                                          {0, 0, 1, 0}, {0, 0, 0, 1},
+                                          {0, 0, 0, 0}, {0, 0, 0, 0}};
+    I<I<value_type>> non_local_restriction = {{0, 0}, {0, 0}, {0, 0},
+                                              {0, 0}, {1, 0}, {0, 1}};
+    I<I<value_type>> local_prolongation = {{1, 0, 0, 0, 0, 0},
+                                           {0, 1, 0, 0, 0, 0},
+                                           {0, 0, 1, 0, 0, 0},
+                                           {0, 0, 0, 1, 0, 0}};
+    I<I<value_type>> non_local_prolongation[] = {
+        {{0}, {0}, {0}, {1}},
+        {{1, 0, 0, 0}, {0, 1, 0, 0}, {0, 0, 1, 0}, {0, 0, 0, 1}},
+        {{1}, {0}, {0}, {0}}};
+
+    this->dist_mat->read_distributed(this->dist_input[rank], this->part);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<csr>(this->dist_mat->get_local_matrix()),
+                        res_local, 0);
+    GKO_ASSERT_MTX_NEAR(
+        gko::as<csr>(this->dist_mat->get_restriction()->get_local_matrix()),
+        local_restriction, 0);
+    GKO_ASSERT_MTX_NEAR(
+        gko::as<csr>(this->dist_mat->get_restriction()->get_non_local_matrix()),
+        non_local_restriction, 0);
+    GKO_ASSERT_MTX_NEAR(
+        gko::as<csr>(this->dist_mat->get_prolongation()->get_local_matrix()),
+        local_prolongation, 0);
+    GKO_ASSERT_MTX_NEAR(
+        gko::as<csr>(
+            this->dist_mat->get_prolongation()->get_non_local_matrix()),
+        non_local_prolongation[rank], 0);
+}
+
+
+TYPED_TEST(DdMatrix, CanApplyToSingleVector)
```

Comment:
Probably also makes sense to add some apply tests to ensure exception throws in cases of dimension mismatches.